### ISSUE TITLE
[PluginEvents] Cypress fires the 'before:spec' event twice

### ISSUE THREAD

**Issue by elaichenkov** [cypress-io/cypress#26300]
### Current behavior

Cypress fires `before:spec` event twice.

```bash
npx cypress run
Missing baseUrl in compilerOptions. tsconfig-paths will be skipped

====================================================================================================

  (Run Starting)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:        12.9.0                                                                         │
  │ Browser:        Electron 106 (headless)                                                        │
  │ Node Version:   v16.15.0 (/Users/ylaichenkov/.nvm/versions/node/v16.15.0/bin/node)             │
  │ Specs:          1 found (todo.cy.js)                                                           │
  │ Searched:       cypress/e2e/**/*.cy.{js,jsx,ts,tsx}                                            │
  │ Experiments:    experimentalInteractiveRunEvents=true                                          │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


────────────────────────────────────────────────────────────────────────────────────────────────────

  Running:  todo.cy.js                                                                      (1 of 1)

--------> Running "before:spec" event <----------



  example to-do app

--------> Running "before:spec" event <----------

    ✓ displays two todo items by default (643ms)


  1 passing (3s)


  (Results)
```
<img width="1006" alt="Screenshot 2023-03-30 at 11 05 51" src="https://user-images.githubusercontent.com/29764053/228771454-f87ddf37-42d8-4c04-8b59-afa15b7f1af6.png">

### Desired behavior

Cypress fires all events only once

### Test code to reproduce

```js
// example.cy.js

describe('example to-do app', () => {
  it('opens todo app', () => {
    cy.visit('https://example.cypress.io/todo')
  })
})
```

```js
// cypress.config.js
const { defineConfig } = require('cypress')

module.exports = defineConfig({
  e2e: {
    setupNodeEvents(on, config) {
      on('before:spec', () => console.log('\n--------> Running "before:spec" event <----------\n'));
    },
  },
})

 ```

### Cypress Version

12.9.0

### Node version

v16.15.0

### Operating System

macOS Version 13.2.1

**Comment by emilyrohrbough** [comment]
I was able to reproduce this behavior and this is not correct behavior. Thank you for the straight-forward reproduction. I actually noticed this same behavior this week.

**Issue:** The reason you are seeing it log 2x is without a `baseUrl` set and your first `cy.visit()` changes top, (or if you do have `baseUrl` set but your visit is to a different domain), Cypress is reloading on the page. 

**Expected Behavior:** When Cypress is reloaded mid-test, it should not re-trigger the `before:spec` lifecycle event.

**Comment by mschile** [comment]
This appears to have been introduced in [v10.8.0](https://docs.cypress.io/guides/references/changelog#10-8-0), possibly with this [commit](https://github.com/cypress-io/cypress/commit/c8b1b303234f8835dadbfc8aeea840f67512ea54).

**Comment by badeball** [comment]
It is true that with v10.7.0, the `before:spec` isn't emitted twice, but task-handlers _are_. IE. the test is re-run for all practical purposes and non-idemponent task handlers must track state somehow. It also means that #26613 isn't strictly a duplicate of this..

**Comment by badeball** [comment]
From a plugin maintainers perspective, this behaviour was actually rather useful, as a test is for all intents and purposes ran again and this allowed such state to be tracked easily. Removing this makes tracking increasingly convoluted.

Edit: from a plugin maintainers perspective, this also represents a major breaking change.
